dm/syncer/validate_worker.go

```diff
@@ -0,0 +1,380 @@
+	rowNotExist
+	rowDifferent
+)
+
+type validateFailedRow struct {
+	tp      validateFailedType
+	dstData []*sql.NullString
+}
+
+type validateWorker struct {
+	cfg               config.ValidatorConfig
+	ctx               context.Context
+	interval          time.Duration
+	validator         *DataValidator
+	L                 log.Logger
+	conn              *dbconn.DBConn
+	rowChangeCh       chan *rowChange
+	pendingChangesMap map[string]*tableChange
+	pendingRowCount   atomic.Int64
+	batchSize         int
+	sync.Mutex
+}
+
+func newValidateWorker(v *DataValidator, id int) *validateWorker {
+	return &validateWorker{
+		cfg:               v.cfg.ValidatorCfg,
+		ctx:               v.ctx,
+		interval:          v.validateInterval,
+		validator:         v,
+		L:                 v.L,
+		conn:              v.toDBConns[id],
+		rowChangeCh:       make(chan *rowChange, workerChannelSize),
+		pendingChangesMap: make(map[string]*tableChange),
+		batchSize:         maxBatchSize,
+	}
+}
+
+func (vw *validateWorker) run() {
+outer:
+	for {
+		select {
+		case change, ok := <-vw.rowChangeCh:
+			if !ok {
+				break outer
+			}
+			// todo: limit number of pending rows.
+			// todo: trigger validation when pending rows count reaches a threshold, but we don't want trigger it too often
+			// todo: since it may not reduce the number of pending row changes as the downstream may have changed again,
+			// todo: we need to catchup with the progress of syncer first, so row changes which are changed again can be merged.
+			vw.updateRowChange(change)
+		case <-time.After(vw.interval):
+			err := vw.validateTableChange()
+			if err != nil {
+				// todo: better error handling
+				vw.validator.errChan <- terror.Annotate(err, "failed to validate table change")
+			}
+			// todo: if a row failed too many times, add it to failed changes
+		}
+	}
+}
+
+func (vw *validateWorker) updateRowChange(row *rowChange) {
+	// cluster using target table
+	fullTableName := row.table.Target.String()
+	change := vw.pendingChangesMap[fullTableName]
+	if change == nil {
+		// no change of this table
+		change = &tableChange{
+			table: row.table,
+			rows:  make(map[string]*rowChange),
+		}
+		vw.pendingChangesMap[fullTableName] = change
+	}
+	if val, ok := change.rows[row.key]; ok {
+		val.data = row.data
+		val.tp = row.tp
+		val.lastMeetTS = row.lastMeetTS
+		val.failedCnt = 0 // clear failed count
+	} else {
+		change.rows[row.key] = row
+		vw.pendingRowCount.Inc()
+	}
+}
+
+func (vw *validateWorker) validateTableChange() error {
+	failedChanges := make(map[string]*tableChange)
+	var failedRowCnt int64
+	for k, tblChange := range vw.pendingChangesMap {
+		var insertUpdateChanges, deleteChanges []*rowChange
+		for _, r := range tblChange.rows {
+			if r.tp == rowDeleted {
+				deleteChanges = append(deleteChanges, r)
+			} else {
+				insertUpdateChanges = append(insertUpdateChanges, r)
+			}
+		}
+		rows := make(map[string]*rowChange)
+		if len(insertUpdateChanges) > 0 {
```

Comment:
seems we can move check `len(deleteChanges) > 0` and `len(insertUpdateChanges) > 0` into func`validateRowChanges `, so logic here can be more simple.